util/filter_bench.cc

```diff
@@ -0,0 +1,403 @@
+
+void _always_assert_fail(int line, const char *file, const char *expr) {
+  fprintf(stderr, "%s: %d: Assertion %s failed\n", file, line, expr);
+  abort();
+}
+
+#define always_assert(cond) \
+  ((cond) ? (void)0 : ::_always_assert_fail(__LINE__, __FILE__, #cond))
+
+using rocksdb::fastrange32;
+using rocksdb::Slice;
+
+struct KeyMaker {
+  KeyMaker(size_t size)
+      : data_(new char[size]),
+        slice_(data_.get(), size),
+        vals_(reinterpret_cast<uint32_t *>(data_.get())) {
+    assert(size >= 8);
+    memset(data_.get(), 0, size);
+  }
+  std::unique_ptr<char[]> data_;
+  Slice slice_;
+  uint32_t *vals_;
+
+  Slice &Get(uint32_t filter_num, uint32_t val_num) {
+    vals_[0] = filter_num + val_num;
+    vals_[1] = val_num;
+    return slice_;
+  }
+};
+
+void PrintWarnings() {
+#if defined(__GNUC__) && !defined(__OPTIMIZE__)
+  fprintf(stdout,
+          "WARNING: Optimization is disabled: benchmarks unnecessarily slow\n");
+#endif
+#ifndef NDEBUG
+  fprintf(stdout,
+          "WARNING: Assertions are enabled; benchmarks unnecessarily slow\n");
+#endif
+}
+
+struct FilterInfo {
+  uint32_t filter_id_ = 0;
+  std::unique_ptr<const char[]> owner_;
+  Slice filter_;
+  uint32_t keys_added_ = 0;
+  std::unique_ptr<rocksdb::FilterBitsReader> reader_;
+  uint64_t outside_queries_ = 0;
+  uint64_t false_positives_ = 0;
+};
+
+enum TestMode {
+  TM_SingleFilter,
+  TM_BatchPrepared,
+  TM_BatchUnprepared,
+  TM_FiftyOneFilter,
+  TM_EightyTwentyFilter,
+  TM_RandomFilter,
+};
+
+static const std::vector<TestMode> allTestModes = {
+    TM_SingleFilter,   TM_BatchPrepared,      TM_BatchUnprepared,
+    TM_FiftyOneFilter, TM_EightyTwentyFilter, TM_RandomFilter,
+};
+
+static const std::vector<TestMode> quickTestModes = {
+    TM_SingleFilter,
+    TM_RandomFilter,
+};
+
+const char *TestModeToString(TestMode tm) {
+  switch (tm) {
+    case TM_SingleFilter:
+      return "Single filter";
+    case TM_BatchPrepared:
+      return "Batched, prepared";
+    case TM_BatchUnprepared:
+      return "Batched, unprepared";
+    case TM_FiftyOneFilter:
+      return "Skewed 50% in 1%";
+    case TM_EightyTwentyFilter:
+      return "Skewed 80% in 20%";
+    case TM_RandomFilter:
+      return "Random filter";
+  }
+  return "Bad TestMode";
+}
+
+struct FilterBench {
+  std::vector<KeyMaker> kms;
+  std::unique_ptr<const rocksdb::FilterPolicy> policy;
+  std::vector<FilterInfo> infos;
+  std::mt19937 random;
+
+  FilterBench()
+      : policy(rocksdb::NewBloomFilterPolicy(FLAGS_bits_per_key)),
+        random(FLAGS_seed) {
+    for (uint32_t i = 0; i < FLAGS_batch_size; ++i) {
+      kms.emplace_back(FLAGS_key_size < 8 ? 8 : FLAGS_key_size);
+    }
+  }
+
+  void Go();
+
+  void RandomQueryTest(bool inside, bool dry_run, TestMode mode);
+};
+
+void FilterBench::Go() {
+  std::unique_ptr<rocksdb::FilterBitsBuilder> builder(
+      policy->GetFilterBitsBuilder());
+
+  uint32_t varianceMask = 1;
```

Comment:
We follow Google C++ Style and the naming convention for variables is like `variance_mask`: https://google.github.io/styleguide/cppguide.html#Variable_Names